### DISCUSSION TITLE
MRIReco.jl Developers/Users at ISMRM

### DISCUSSION THREAD

**Discussion by alexjaffray** [MagneticResonanceImaging/MRIReco.jl#82] [General]
Hi all, 

I think the upcoming ISMRM might be a good chance for some of the people who develop and/or use this project to meet in-person and discuss ideas. @mrikasper and I would be happy to share our experiences using the package in practice, and learn from what other people have done with it as well!

Thoughts?

**Comment by tknopp** [comment]
I will unfortunately not be at ISMRM but would of course encourage you to meet and discuss ideas.

**Comment by aTrotier** [comment]
I will be at ISMRM, we can setup a meeting.

I don't have a lot of things to show because I am still mostly using BART for the CS processing but I'll be happy to discuss about the way to finish that transition.

By the way we can also have a discussion about the creation of an online meeting on a regular basis like the one in [gadgetron](https://github.com/gadgetron/gadgetron/discussions/1093) ?

**Reply by alexjaffray** [reply]
Let's do that. I am also interested in CS recons in Julia so would be very happy to ask some questions about your experience and see what is needed to move forward.

I think a regular meeting is a good idea, and a starting point for a more structured way forward for the package. Let's definitely discuss it when we meet at the ISMRM.

**Reply by mrikasper** [reply]
Dear @aTrotier and @alexjaffray,

I will also be at ISMRM and would be interested to meet with MRIReco.jl users. Not sure whether @JakobAsslaender and @JeffFessler have seen this discussion, they might also be around and interested to join?

Just as a first iteration for finding a time and place, @alexjaffray's [poster](http://submissions.mirasmart.com/ISMRM2022/ViewSubmissionFile.aspx?sbmID=1857&validate=false&mode=html&key=Brm71oQ88) uses MRIReco, so maybe we could meet [there (Wed, 17:45-18:45 UTC)](https://ismrm-esmrmb-ismrt2022.us3.pathable.com/meetings/virtual/poster/fMF7RmrsyKKjWbiYN) or before/after for a beverage and chat? 

I also tried to find other Julia/MRIReco abstracts/educationals at ISMRM that would be obvious gathering points, but didn't find anything meaningful. Maybe the keyword "Julia" is too ambiguous :-)

All the best,
Lars

**Reply by JeffFessler** [reply]
I will plan to come to that poster, and I may bring some group members - thanks for the invite!

**Reply by aTrotier** [reply]
Seems good to me :)

**Reply by alexjaffray** [reply]
Fantastic. I am looking forward to it.

**Comment by alexjaffray** [comment]
It was great to see everyone at our little informal meeting at the ISMRM, it was exciting to hear about what people are working on in Julia! 

I think that the big takeaway from the gathering was that it would be useful to set up a recurring meeting for the developers/users of the package. This way we could, as a community, define a roadmap forward for the package and other tools in the broader Julia MR ecosystem. 

@tknopp, @migrosser, @mrikasper, @JeffFessler, @JakobAsslaender, @aTrotier: let me know what you think, and how we should possibly schedule such a meeting.

**Reply by aTrotier** [reply]
Good idea !
The gadgetron community has a meeting each Friday but they have people dedicating to its development. It seems more reasonable to start with a meeting each month or 2 ?

It will be great to define a roadmap and maybe to define priorities linked to our common research topics ?
It could also be a good opportunities as a user to discuss our current issues with the package and get some helps which then can be converted to a tutorial / documention enhancement.

**Reply by mrikasper** [reply]
Dear MRIReco-nnoisseurs,

It was really great to see so many Julia enthusiasts at ISMRM and exchange ideas. I would also be very interested to meet up regularly, and also understand a bit better what vision @tknopp and @migrosser have for MRIReco.jl, so that we can identify synergies, common values, but also our different use cases, in order to expand the use of Julia and MRIReco.jl in our MRI Recon community.

All the best,
Lars

**Reply by aTrotier** [reply]
Not a lot of possibility for a recurring meeting between vancouver and Hamburg -> Between 3:00pm  to 4:00pm - GMT is one the the slots available.

**Reply by alexjaffray** [reply]
If the meeting is biweekly (or monthly) then I don't mind waking up early so that it is more convenient for people in Europe. I think I would be the only west coast person currently, so probably finding a time that would work between Toronto/Eastern USA and Europe is easier to start with.

**Reply by tknopp** [reply]
Sorry for being silent here, for some reasons I didn't get email notifications by Github anymore. But I have fixed this and hope that I can be more reactive from now on.

Mirco and I are currently quite busy with other projects during the summer and will not really have time to work on larger things within MRIReco.jl. So, we can certainly setup a virtual meeting and we try to participate but I don't want to organize this myself.

**Reply by alexjaffray** [reply]
No problem @tknopp, that's alright! 

I think that we can try to find a time for now that is reasonable between Eastern Standard Time and Central European Time, and collaborators can join if possible or propose an alternate time if it ends up being problematic. My schedule is rather flexible, so some time around/after 2pm CET would be a good starting point. Thoughts @aTrotier?

**Reply by JeffFessler** [reply]
2pm CET is 8am ET and I'm typically available then (when not traveling). I'm happy to give the 1/month sampling rate a try.

**Reply by aTrotier** [reply]
Let's try that. I am available in July for a meeting but it might be better to start after the holidays ? Early september ?

**Reply by alexjaffray** [reply]
@aTrotier I am also available in July for a meeting, perhaps we can start with something like the last week of July or so, or September as you suggest

**Reply by mrikasper** [reply]
I would also be available for end of July or early September at that time.

**Reply by monreal93** [reply]
Hi all,
I would also like to join the meetings, I am available at the end of July and September as well!

**Reply by aTrotier** [reply]
**First thursday of each month -> 2pm CET** 
Let us know if it is a problem for one of you ?

If it is ok, we schedule a virtual meeting **September 1st**.

**Reply by cncastillo** [reply]
Hi! I would love to be in the meetings as well, 2 pm CET on the 1st of September is ok with me :smile:. I am the developer of [KomaMRI.jl](https://github.com/cncastillo/KomaMRI.jl) an MRI simulator fully written in Julia (besides the UI). Currently using MRIReco.jl for the reconstructions (as BART would do in JEMRIS).

**Comment by aTrotier** [comment]
The first meeting will be held : 
**September 1st -> 2pm CET**

Do you have some topics you want to discuss ?

A list with some ideas (some of them might not be relevant):
>## About MRIReco :
>- How we use it 
>- Splitting package : feedback
>- enhancement / future plans ?
>- Benchmark reconstruction time (example : reconstruction functions vs building the operator by hand)
>- Comparison of CS results with BART : https://atrotier.github.io/EDUC_JULIA_CS_MP2RAGE/
>
>## MRI Community :
>- EPG package ? @jondeuce
>- Dedicated MRI viewer (missing features from ImageViewer.jl )

**Reply by mrikasper** [reply]
- Development Philosophy
    - Coding/Contributing guidelines (e.g., naming conventions, interface requirements, e.g., for types)
    - Documentation (see [this issue](https://github.com/MagneticResonanceImaging/MRIReco.jl/issues/7)), in particular for (optional) arguments and their valid values
    - Inclusion of external contributors' use cases (e.g., creating specific tests to ensure updates don't break users' workflow)
    - How to identify synergies in development with external contributors, i.e., splitting the work for commonly desired new(? haven't checked lately) features, e.g., 
        - [higher order phase recon](https://doi.org/10.1002/mrm.22767)
        - GPU support
        - arbitrary 3D trajectory recon

**Reply by aTrotier** [reply]
> arbitrary 3D trajectory recon

That's already the case, no ?

**Reply by mrikasper** [reply]
That would be great. I checked about 2y ago, and then for the iterative recon with gridding, time-segmentation and B0 correction (cg-SENSE), I seem to remember that the 3rd dimension was just an FFT, but I might be mistaken.

**Reply by mrikasper** [reply]
Maybe it's just the documentation that tripped me : https://github.com/MagneticResonanceImaging/MRIReco.jl/blob/aaff63fe8aaf1ea4092df5abb786139be0ae9789/src/Reconstruction/IterativeReconstruction.jl#L209

Is there a 3D example for arbitrary, B0-corrected SENSE recon in the documentation? If not, then I would opt to discuss having one (and do the work :-))

**Reply by JakobAsslaender** [reply]
I converted a few algorithms to 3D (ND to be more precise) and so did @tknopp and others, but this process is certainly far from complete and I am sure that help is more than welcome! The high level tools in MRIReco.jl are certainly a personal blind spot of mine as I am not using them, but NFFT.jl, ESPRIT, SparsityOperators.jl, ReguarlizedLeastSquares.jl etc. should all work for 3D imaging.

**Reply by aTrotier** [reply]
That's the idea of the meeting, If you have real needs for the high level tools we can decide to implement them (and build examples / tests).

For my research, I am mostly using the conversion to RAW / acquistionData and then i'm using the low-level function of MRIReco to perform sense estimation, FT or nufft operator, density compensation.

**Reply by mrikasper** [reply]
That's great to hear, thank you for the pointers. I also have a lot of blind spots with respect to the Julia ecosystem around MRIReco. 

And yes, I would be keen to invest more into high level workflows, tests, and documentation. SMS or 3D stack-of-spirals (tilted hexagonal spirals, to be precise :-)) would be areas of application we have in mind.

**Reply by aTrotier** [reply]
**Additionnal discussion :** it would be great to have some kind of consensus about our different packages for the dimension position. 

I think MRIReco is based on a 5D array with : x,y,z,echo,channel,NR but some of the function like espirit  returns array with other dimensions. 

For information BART uses the convention :
```C
enum mri_dims {
	READ_DIM,
	PHS1_DIM,
	PHS2_DIM,
	COIL_DIM,
	MAPS_DIM,
	TE_DIM,
	COEFF_DIM,
	COEFF2_DIM,
	ITER_DIM,
	CSHIFT_DIM,
	TIME_DIM,
	TIME2_DIM,
	LEVEL_DIM,
	SLICE_DIM,
	AVG_DIM,
	BATCH_DIM,
};
```

**Reply by mrikasper** [reply]
In principle, I see two ways to go about this: flexible dimensionality with agreed-upon labels (and a label vector, `AxisArrays` does this, right?) or a fixed high-dimensional data array that all interfaces respect. I find the former solution more elegant (and would assume it's more performant), but the established packages and data formats (e.g., ISMRMRD) I know of in Matlab or C seem to prefer the latter...I guess because a fixed convention is easier to maintain than flexible interfacing. But I don't know whether this is also true for Julia.

**Reply by JakobAsslaender** [reply]
My guess is that C and Matlab do not have this functionality build in, do they? I would vote for using `AxisArrays` instead of a high-dimensional array as the latter are very unintuitive. To me, the more important question is where to include this. For the more basic code/packages, I do not think that we need it and I would refrain from making the code more MRI specific than necessary. I would rather include such things only in high-level wrapper packages like `MRIReco.jl`.

**Comment by JeffFessler** [comment]
Do we need a zoom URL for the meeting next week?

**Reply by aTrotier** [reply]
Yes. I can create one but it would be better if someone else is the admin. 
(1st day of school for my daughter, she might be around during the meeting...)

**Comment by migrosser** [comment]
I took the freedom to create a zoom-link for our meeting. Looking forward to seeing you tomorrow.
https://tuhh.zoom.us/j/82596367645?pwd=Zld0NDRXbitPMCt5QzBsSS9zYVNoUT09

**Reply by JakobAsslaender** [reply]
Sorry, I won't make it to the meeting tomorrow.

**Comment by mrikasper** [comment]
Hi everyone,

as discussed today, I started our MRIReco Community Meetings discussion in a new thread: https://github.com/MagneticResonanceImaging/MRIReco.jl/discussions/99

Please have a look, I also put some minutes to review and amend.

**Comment by aTrotier** [comment]
## Topics for next meeting 
_Thu, Oct, 6th at 12:00 UTC_

- Benchmark results
- Accuracy of fista (example in benchmark or https://atrotier.github.io/EDUC_JULIA_CS_MP2RAGE/)
- Reconstruction of repetitions : PR #105 
- Conversion from acq to BartIO and BartIO to acq ? (also part of PR #105)
- Be able to use different trajectories for all dimensions (only implemented for echoes)
- ISMRM abstracts ?

**Reply by mrikasper** [reply]
I just saw you added topics here, @aTrotier, was wondering whether we should consolidate it in #99

**Reply by aTrotier** [reply]
Let's do that
